cmd/maya-apiserver/cstor-operator/cspc/handler.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/pkg/errors"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	// supportedPool is a map holding the supported raid configurations.
+	supportedPool = map[apis.CasPoolValString]bool{
+		apis.PoolTypeStripedCPV:  true,
+		apis.PoolTypeMirroredCPV: true,
+		apis.PoolTypeRaidzCPV:    true,
+		apis.PoolTypeRaidz2CPV:   true,
+	}
+)
+
+type clientSet struct {
+	oecs openebs.Interface
+}
+
+// PoolConfig embeds nodeselect config from algorithm package and Controller object.
+type PoolConfig struct {
+	AlgorithConfig *nodeselect.Config
+	Controller     *Controller
+}
+
+// NewPoolConfig returns a poolconfig object
+func (c *Controller) NewPoolConfig(cspc *apis.CStorPoolCluster, namespace string) (*PoolConfig, error) {
+	pc, err := nodeselect.
+		NewBuilder().
+		WithCSPC(cspc).
+		WithNameSpace(namespace).
+		Build()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get algorithm config for provisioning")
+	}
+	return &PoolConfig{AlgorithConfig: pc, Controller: c}, nil
+
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the cspcPoolUpdated resource
+// with the current status of the resource.
+func (c *Controller) syncHandler(key string) error {
+	startTime := time.Now()
+	glog.V(4).Infof("Started syncing storagepoolclaim %q (%v)", key, startTime)
+	defer func() {
+		glog.V(4).Infof("Finished syncing storagepoolclaim %q (%v)", key, time.Since(startTime))
+	}()
+
+	// Convert the namespace/name string into a distinct namespace and name
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the cspc resource with this namespace/name
+	cspc, err := c.cspcLister.CStorPoolClusters(ns).Get(name)
+	if k8serror.IsNotFound(err) {
+		runtime.HandleError(fmt.Errorf("cspc '%s' has been deleted", key))
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Deep-copy otherwise we are mutating our cache.
+	// TODO: Deep-copy only when needed.
+	cspcGot := cspc.DeepCopy()
+	err = c.syncCSPC(cspcGot)
+	return err
+}
+
+// enqueueSpc takes a SPC resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than SPC.
+func (c *Controller) enqueueSpc(cspc interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(cspc); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// synSpc is the function which tries to converge to a desired state for the cspc.
+func (c *Controller) syncCSPC(cspc *apis.CStorPoolCluster) error {
+	//err := validate(cspc)
+	//if err != nil {
+	//	glog.Errorf("Validation of cspc failed:%s", err)
+	//	return nil
+	//}
+	openebsNameSpace := env.Get(env.OpenEBSNamespace)
```

Comment:
are we handling delete event?

Reply:
It is already handled via owner references although I will run one round of sanity test to verify this.